File: slimadapter/src/main/java/com/jhj/slimadapter/adapter/DraggableAdapter.java
package com.jhj.slimadapter.adapter;

import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

import com.jhj.slimadapter.itemtouch.ItemTouchHelperCallback;
import com.jhj.slimadapter.listener.OnItemDragListener;
import com.jhj.slimadapter.listener.OnItemSwipeListener;

/**
 * 在　SlimAdapter 的基础上可以实现长按拖动，滑动删除（除了标题和结尾外的的布局）
 * <p>
 * Created by jhj on 18-10-23.
 */

public class DraggableAdapter extends BaseAdapter<DraggableAdapter> {

    private boolean itemDragEnabled = false;
    private boolean itemSwipeEnabled = false;
    private ItemTouchHelperCallback itemTouchHelperCallback;


    private DraggableAdapter(RecyclerView.LayoutManager manager) {
        this.layoutManager = manager;
    }

    public static DraggableAdapter creator(RecyclerView.LayoutManager manager) {
        return new DraggableAdapter(manager);
    }

    public DraggableAdapter setItemTouchHelper() {
        itemTouchHelperCallback = new ItemTouchHelperCallback(this);
        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(itemTouchHelperCallback);
        itemTouchHelper.attachToRecyclerView(getRecyclerView());
        return this;
    }

    //======　drag　======


    public DraggableAdapter setDragItem(boolean isDrag) {
        this.itemDragEnabled = isDrag;
        return this;
    }

    public DraggableAdapter setDragFlags(int dragFlags) {
        itemTouchHelperCallback.setDragFlags(dragFlags);
        return this;
    }

    public DraggableAdapter setOnItemDragListener(OnItemDragListener onItemDragListener) {
        itemTouchHelperCallback.setOnItemDragListener(onItemDragListener);
        return this;
    }

    public DraggableAdapter setMoveThreshold(float moveThreshold) {
        itemTouchHelperCallback.setMoveThreshold(moveThreshold);
        return this;
    }

    public boolean isLongPressDragEnable() {
        return itemDragEnabled;
    }

    //======= swipe =======


    public DraggableAdapter setSwipeItem(boolean isSwipe) {
        itemSwipeEnabled = isSwipe;
        return this;
    }


    public DraggableAdapter setSwipeFlags(int swipeFlags) {
        itemTouchHelperCallback.setSwipeFlags(swipeFlags);
        return this;
    }

    public DraggableAdapter setOnItemSwipeListener(OnItemSwipeListener listener) {
        itemTouchHelperCallback.setOnItemSwipeListener(listener);
        return this;
    }

    public DraggableAdapter setSwipeThreshold(float swipeThreshold) {
        itemTouchHelperCallback.setSwipeThreshold(swipeThreshold);
        return this;
    }

    public DraggableAdapter setSwipeFadeOutAnim(boolean swipeFadeOutAnim) {
        itemTouchHelperCallback.setSwipeFadeOutAnim(swipeFadeOutAnim);
        return this;
    }

    public boolean isItemSwipeEnable() {
        return itemSwipeEnabled;
    }

}

File: slimadapterkt/src/main/java/com/zgdj/slimadapterkt/DraggableAdapter.kt
package com.zgdj.slimadapterkt

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper

import com.zgdj.slimadapterkt.itemtouch.ItemTouchHelperCallback
import com.zgdj.slimadapterkt.listener.OnItemDragListener
import com.zgdj.slimadapterkt.listener.OnItemSwipeListener

/**
 * 在　SlimAdapter 的基础上可以实现长按拖动，滑动删除（除了标题和结尾外的的布局）
 *
 *
 * Created by jhj on 18-10-23.
 */

class DraggableAdapter private constructor(manager: RecyclerView.LayoutManager) : BaseAdapter<DraggableAdapter>() {

    var isLongPressDragEnable = false
        private set
    var isItemSwipeEnable = false
        private set
    private var itemTouchHelperCallback: ItemTouchHelperCallback? = null


    init {
        this.layoutManager = manager
    }

    fun setItemTouchHelper(): DraggableAdapter {
        itemTouchHelperCallback = ItemTouchHelperCallback(this)
        val itemTouchHelper = ItemTouchHelper(itemTouchHelperCallback!!)
        itemTouchHelper.attachToRecyclerView(getRecyclerView())
        return this
    }

    //======　drag　======


    fun setDragItem(isDrag: Boolean): DraggableAdapter {
        this.isLongPressDragEnable = isDrag
        return this
    }

    fun setDragFlags(dragFlags: Int): DraggableAdapter {
        itemTouchHelperCallback!!.setDragFlags(dragFlags)
        return this
    }

    fun setOnItemDragListener(onItemDragListener: OnItemDragListener): DraggableAdapter {
        itemTouchHelperCallback!!.setOnItemDragListener(onItemDragListener)
        return this
    }

    fun setMoveThreshold(moveThreshold: Float): DraggableAdapter {
        itemTouchHelperCallback!!.setMoveThreshold(moveThreshold)
        return this
    }

    //======= swipe =======


    fun setSwipeItem(isSwipe: Boolean): DraggableAdapter {
        isItemSwipeEnable = isSwipe
        return this
    }


    fun setSwipeFlags(swipeFlags: Int): DraggableAdapter {
        itemTouchHelperCallback!!.setSwipeFlags(swipeFlags)
        return this
    }

    fun setOnItemSwipeListener(listener: OnItemSwipeListener): DraggableAdapter {
        itemTouchHelperCallback!!.setOnItemSwipeListener(listener)
        return this
    }

    fun setSwipeThreshold(swipeThreshold: Float): DraggableAdapter {
        itemTouchHelperCallback!!.setSwipeThreshold(swipeThreshold)
        return this
    }

    fun setSwipeFadeOutAnim(swipeFadeOutAnim: Boolean): DraggableAdapter {
        itemTouchHelperCallback!!.setSwipeFadeOutAnim(swipeFadeOutAnim)
        return this
    }

    companion object {

        fun creator(manager: RecyclerView.LayoutManager): DraggableAdapter {
            return DraggableAdapter(manager)
        }
    }

}
